### PR TITLE
Lwdev 3235 erc20 private wallet

### PR DESCRIPTION
**Pull request by OttoVT**
https://lykkex.atlassian.net/browse/LWDEV-3235

### CODE REVIEW

File: src/EthereumApi/Controllers/Erc20BalanceController.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        {
+            _erc20BalanceService = erc20BalanceService;
+            _log = log;
+        }
+
+        [HttpPost]
+        [ProducesResponseType(typeof(AddressTokenBalanceContainerResponse), 200)]
+        [ProducesResponseType(typeof(ApiException), 400)]
+        [ProducesResponseType(typeof(ApiException), 500)]
+        public async Task<IActionResult> GetTransaction([FromBody]GetErcBalance ercTransaction)
```
Comment (paradise) [inline src/EthereumApi/Controllers/Erc20BalanceController.cs]: why "GetTransaction" and "ercTransaction", if it about balance ?

File: src/Service.UnitTests/Config.cs [hunk 2]
```diff
@@ -26,6 +27,9 @@ public class Config
             IServiceCollection collection = new ServiceCollection();
             collection.AddSingleton<INonceCalculator, Mocks.MockNonceCalculator>();
             collection.AddSingleton<IEthereumSamuraiApi, Mocks.MockEthereumSamuraiApi>();
+            collection.AddSingleton<IEthereumSamuraiApi, Mocks.MockEthereumSamuraiApi>();
+            collection.AddSingleton<ISignatureChecker, SignatureChecker>();
+            //collection.AddSingleton<IRawTransactionSubmitter, RawTransactionSubmitter>();
```
Comment (paradise) [inline src/Service.UnitTests/Config.cs]: maybe it can be removed

File: src/Services/Coins/ExchangeContractService.cs [hunk 3]
```diff
@@ -304,6 +307,31 @@ public bool IsValidAddress(string address)
...
+
+            var contract = _web3.Eth.GetContract(_settings.MainExchangeContract.Abi, _settings.MainExchangeContract.Address);
+            var transferFunction = contract.GetFunction("transfer");
+
+            var convertedId = EthUtils.GuidToBigInteger(id);
+            var transactionHash = await transferFunction.SendTransactionAsync(_settings.EthereumMainAccount,
+                    new HexBigInteger(Constants.GasForCoinTransaction), new HexBigInteger(0),
+                    convertedId, coinAFromDb.AdapterAddress, from, to, amount, sign.HexToByteArray().FixByteOrder(), new byte[0]);
+            await SaveUserHistory(coinAddress, amount.ToString(), from, to, transactionHash, "Transfer");
+            //await _coinEventService.PublishEvent(new CoinEvent(id.ToString(), transactionHash, from, to,
```
Comment (paradise) [inline src/Services/Coins/ExchangeContractService.cs]: comments

File: src/Services/Coins/ExchangeContractService.cs [hunk 4]
```diff
@@ -407,6 +435,19 @@ public bool CheckSign(Guid id, string coinAddress, string clientAddr, string toA
...
         }
 
+        public async Task<string> ChangeMainContractInCoin(string coinAddress, string newExchangeContractAddress, string newMainExchangeAbi)
+        {
+            var coinAFromDb = await GetCoinWithCheck(coinAddress);
+            var contract = _web3.Eth.GetContract(_settings.MainExchangeContract.Abi, _settings.MainExchangeContract.Address);
+            var transferFunction = contract.GetFunction("changeMainContractInCoin");
+            var transactionHash = await transferFunction.SendTransactionAsync(_settings.EthereumMainAccount,
+                    new HexBigInteger(Constants.GasForCoinTransaction), new HexBigInteger(0), coinAddress, newExchangeContractAddress);
+            //_settings.MainExchangeContract.Abi = newMainExchangeAbi;
```
Comment (paradise) [inline src/Services/Coins/ExchangeContractService.cs]: comments

File: src/Services/New/PendingOperationService.cs [hunk 5]
```diff
@@ -436,5 +462,22 @@ private byte[] GetHash(Guid id, string coinAddress, string clientAddr, string to
             await _log.WriteInfoAsync("PendingOperationService", "GetSign", "", $"ID({id})-COINADDRESS({coinAddress})-FROM({clientAddr})-TO({toAddr})-AMOUNT({amount})-HASH({response.SignedHash})", DateTime.UtcNow);
             return response.SignedHash;
         }
+
+        public async Task RemoveFromPendingOperationQueue(string opId)
+        {
+            var count = await _queue.Count();
+            for (int i = 0; i < count; i++)
+            {
+                var message = _queue.GetRawMessageAsync().Result;
```
Comment (paradise) [inline src/Services/New/PendingOperationService.cs]: await, not Result

File: src/Services/New/PendingOperationService.cs [hunk 6]
```diff
@@ -436,5 +462,22 @@ private byte[] GetHash(Guid id, string coinAddress, string clientAddr, string to
...
+
+                OperationHashMatchMessage newMessage = JsonConvert.DeserializeObject<OperationHashMatchMessage>(message.AsString);
+
+                await _queue.FinishRawMessageAsync(message);
+                if (newMessage.OperationId?.ToLower() != opId?.ToLower())
+                {
+                    _queue.PutRawMessageAsync(message.AsString).Wait();
```
Comment (paradise) [inline src/Services/New/PendingOperationService.cs]: await

File: src/Services/PrivateWallet/Erc20Service.cs [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+            await ValidateTokenAddressBalanceAsync(transaction.FromAddress, transaction.TokenAddress, transaction.TokenAmount);
+        }
+
+        public async Task ValidateInputForSignedAsync(string fromAddress, string signedTransaction)
+        {
+            await _transactionValidationService.ValidateInputForSignedAsync(fromAddress, signedTransaction);
+            Nethereum.Signer.Transaction transaction = new Nethereum.Signer.Transaction(signedTransaction.HexToByteArray());
+            string erc20Address                      = transaction.ReceiveAddress.ToHexCompact().EnsureHexPrefix();
+            string erc20InvocationData               = transaction.Data.ToHexCompact().EnsureHexPrefix();
+            //0xa9059cbb000000000000000000000000aa4981d084120aef4bbaeecb9abdbc7d180c7edb000000000000000000000000000000000000000000000000000000000000000a
```
Comment (paradise) [inline src/Services/PrivateWallet/Erc20Service.cs]: comments

File: src/Services/PrivateWallet/Erc20Service.cs [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+            if (! await _transactionValidationService.IsTransactionErc20Transfer(signedTransaction))
+            {
+                throw new ClientSideException(ExceptionType.WrongParams, "Transaction is not a erc20 transfer");
+            }
+
+            string parametrsString    = erc20InvocationData.Replace(Constants.Erc20TransferSignature, "");
+            var amount                = parametrsString.Substring(64, 64);
+            //string addressString    = parametrsString.Substring(0, 64).TrimStart(new char[] { '0' }).EnsureHexPrefix();
```
Comment (paradise) [inline src/Services/PrivateWallet/Erc20Service.cs]: comments

File: src/Services/Transactions/RawTransactionSubmitter.cs [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            var ethSendTransaction = new EthSendRawTransaction(_web3.Client);
+            string transactionHex;
+
+            try
+            {
+                transactionHex = await ethSendTransaction.SendRequestAsync(signedTrHex);
+            }
+            catch (Nethereum.JsonRpc.Client.RpcResponseException ex)
+            {
+                if (ex.Message == "intrinsic gas too low")
```
Comment (paradise) [inline src/Services/Transactions/RawTransactionSubmitter.cs]: maybe you can use .RpcError.Code ?

File: src/EthereumApi/Controllers/Erc20BalanceController.cs (hunk 1 above)
Comment (OttoVT) [reply src/EthereumApi/Controllers/Erc20BalanceController.cs]: fixed

File: src/Service.UnitTests/Config.cs (hunk 2 above)
Comment (OttoVT) [reply src/Service.UnitTests/Config.cs]: fixed

File: src/Services/Coins/ExchangeContractService.cs (hunk 3 above)
Comment (OttoVT) [reply src/Services/Coins/ExchangeContractService.cs]: fixed

File: src/Services/Coins/ExchangeContractService.cs (hunk 4 above)
Comment (OttoVT) [reply src/Services/Coins/ExchangeContractService.cs]: fixed

File: src/Services/New/PendingOperationService.cs (hunk 5 above)
Comment (OttoVT) [reply src/Services/New/PendingOperationService.cs]: fixed

File: src/Services/New/PendingOperationService.cs (hunk 6 above)
Comment (OttoVT) [reply src/Services/New/PendingOperationService.cs]: fixed

File: src/Services/PrivateWallet/Erc20Service.cs (hunk 7 above)
Comment (OttoVT) [reply src/Services/PrivateWallet/Erc20Service.cs]: need that for newcomers

File: src/Services/PrivateWallet/Erc20Service.cs (hunk 8 above)
Comment (OttoVT) [reply src/Services/PrivateWallet/Erc20Service.cs]: fixed

File: src/Services/Transactions/RawTransactionSubmitter.cs (hunk 9 above)
Comment (OttoVT) [reply src/Services/Transactions/RawTransactionSubmitter.cs]: No, I can't. Error code for such response is 32000 as for many other error types

**Pull request merged**